packages/flutter_tools/lib/src/ios/ios_deploy.dart

```diff
@@ -169,30 +211,193 @@ class IOSDeploy {
+        _iosDeployEnv = iosDeployEnv,
+        _debuggerOutput = debuggerOutput ?? StreamController<String>.broadcast();
+
+  factory IOSDeployDebugger.test({
+    ProcessManager processManager,
+    Logger logger,
+  }) {
+    final Logger debugLogger = logger ?? BufferLogger.test();
+    return IOSDeployDebugger(
+      logger: debugLogger,
+      processUtils: ProcessUtils(logger: debugLogger, processManager: processManager),
+      launchCommand: <String>['ios-deploy'],
+      iosDeployEnv: <String, String>{},
+    );
+  }
+
+  final Logger _logger;
+  final ProcessUtils _processUtils;
+  final List<String> _launchCommand;
+  final Map<String, String> _iosDeployEnv;
+
+  Process _iosDeployProcess;
+
+  Stream<String> get logLines => _debuggerOutput.stream;
+  final StreamController<String> _debuggerOutput;
+
+  bool get debuggerAttached => _debuggerAttached;
+  bool _debuggerAttached = false;
+
+  /// Launch the app on the device, and attach the debugger.
+  ///
+  /// Returns whether or not the debugger successfully attached.
+  Future<bool> launchAndAttach() async {
+    // Return when the debugger attaches, or the ios-deploy process exits.
+    final Completer<bool> debuggerCompleter = Completer<bool>();
+    try {
+      _iosDeployProcess = await _processUtils.start(
+        _launchCommand,
+        environment: _iosDeployEnv,
+      );
+      bool lldbRunCalled = false;
+      String lastLineFromDebugger;
+      final StreamSubscription<String> stdoutSubscription = _iosDeployProcess.stdout
+          .transform<String>(utf8.decoder)
+          .transform<String>(const LineSplitter())
+          .listen((String line) {
+        _monitorIOSDeployFailure(line, _logger);
+
+        // (lldb)     run
+        // success
+        // 2020-09-15 13:42:25.185474-0700 Runner[477:181141] flutter: Observatory listening on http://127.0.0.1:57782/
+        if (line == '(lldb)     run') {
+          _logger.printTrace(line);
+          lldbRunCalled = true;
+          return;
+        }
+        // next line must be "success".
+        if (lldbRunCalled && line == 'success') {
+          _logger.printTrace(line);
+          _debuggerAttached = true;
+          debuggerCompleter.complete(true);
+          return;
+        }
+        lldbRunCalled = false;
+        if (line.contains('PROCESS_STOPPED') ||
+            line.contains('PROCESS_EXITED')) {
+          // The app exited or crashed, so stop echoing the output.
+          // Don't pass any further ios-deploy debugging messages to the log reader after it exits.
+          _debuggerAttached = false;
+          _logger.printTrace(line);
+          return;
+        }
+
+        if (line.contains('error: process launch failed')) {
+          // Timed out waiting for the app to launch.
+          _debuggerAttached = false;
+          _logger.printTrace(line);
+          debuggerCompleter.complete(false);
+          return;
+        }
+        if (!debuggerAttached) {
```

Comment:
Oh, okay. The very next line after the lldb run line doesn't have to be 'success'. We'll wait until it is. Consider updating the comment above.

Reply:
It does have to be "success" (`lldbRunCalled ` logic where it's being set to false if the next line isn't "success"), unless I have a bug in here.

Maybe I should change this to fail if the next line isn't "success" so it doesn't hang forever on something unexpected instead of special-casing "error: process launch failed" (and example of a non-success next line).

Reply:
Ah, okay. I misread the code. My suggestion for a cleanup would be to structure this explicitly as a state machine, with an enum setting out explicitly what the states are, and a big comment explaining what/when the transitions are.

That might also make testing this code easier. I think I see below that the error paths in here could use some coverage.

Reply:
Added an enum, hopefully it's clearer now.